### ISSUE TITLE
Issue with ThumbnailerImageField (ImageField?)

### ISSUE THREAD

**Issue by hauru** [bruth/django-preserialize#13]
Hi,

I'm using [Easy Thumbnails](https://github.com/SmileyChris/easy-thumbnails) for managing resized versions of images. In my model i have a [ThumbnailerImageField](https://github.com/SmileyChris/easy-thumbnails/blob/master/easy_thumbnails/fields.py#L32) which stores a URL to the original image. When trying to serialize an instance of this model, i get the JSON exception:

`<ThumbnailerImageFieldFile: ...> is not JSON serializable`

`ThumbnailerImageField` extends Django's original `ImageField` and the Django-provided serialization mechanism works well in this case so i guess it may be a bug.

**Comment by bruth** [comment]
Hi @hauru. preserialize does not actually encode the data into JSON. It simply creates a data structure based on the arguments in the [serialize function](https://github.com/bruth/django-preserialize/blob/master/preserialize/serialize.py#L149). So, if the types in the data structure cannot be encoded by the Django JSON encoder (and therefore the standard library decoder), then the encode may need to be extended to support this type.

**Comment by hauru** [comment]
Yeah, sorry for that. Actually i know it isn't supposed to serialize into JSON but it was late yesterday when i was testing it and, as it turns out, my mind was far from being clear.

What i did was to drop-in replace my own serializer, which also converts model instances to regular Python dictionaries. Resulting data is JSON-encoded later with a different function call (using Django's own encoder). I guess what i was expecting from preserialize was the output dictionary values to be plain Python objects. In case of an `ImageField` it would mean a regular string with URL. This way doing actual serialization (JSON, XML etc.) wouldn't involve dealing with model fields anymore.

**Comment by bruth** [comment]
No problem at all.

>  I guess what i was expecting from preserialize was the output dictionary values to be plain Python objects.

I understand the expectation. However, since arbitrary fields could be defined on a model it would be impossible to account for all cases. That being said, a possible approach would be to look coerce objects to a primitive based on it's internal type.

One way of handling this right now, is to use a [`posthook`](https://github.com/bruth/django-preserialize#hooks) which take the model instance and the serialized attributes to perform some processing, such as getting the path string for the image file.

**Comment by bruth** [comment]
The coercion could use the [`get_prep_value`](https://github.com/django/django/blob/master/django/db/models/fields/__init__.py#L706-L712) method on fields. For example, here is [`FileField`](https://github.com/django/django/blob/master/django/db/models/fields/files.py#L284-L290) and it returns a string.

**Comment by hauru** [comment]
That's what i eventually end up doing. Well, more or less.

In my case the difficulty with `posthook` is that it has to be provided separately for the parent model and for each of the "related" fields. It's inconvenient since i need the same common logic to be applied for all model instances (i have `ImageField`s in many interrelated models). I ended up modifying your code and adding a separate "global" posthook which gets propagated from the parent `serialize` call to subsequent calls for instance's querysets.

Thanks once again!

**Comment by bruth** [comment]
I have a couple of improvements in the works. #14 defines the `Serializer` class to encapsulate these options rather than needing to use globals. I also have a [branch](https://github.com/bruth/django-preserialize/commit/094dbd13ccc8f87eb3912d9df1fdb6048e35d25d) to use `get_prep_value` as discussed above.

On a side note, some of the CI tests are failing due to Django's [`AppRegistryNotReady` error](https://travis-ci.org/bruth/django-preserialize/jobs/111288054#L183). I am less familiar with Django 1.7+ Any thoughts on this? The [test script](https://github.com/bruth/django-preserialize/blob/master/test_suite.py) and [tests package](https://github.com/bruth/django-preserialize/tree/master/tests) are pretty basic.

**Comment by bruth** [comment]
Looks like I figured it out. I was not calling `django.setup()` for versions greater than 1.7.
